File: backend/app/schemas/tournament.py
from pydantic import BaseModel, ConfigDict, field_validator, Field
from typing import Optional, Literal, Any
from datetime import datetime


# ── Participant info ───────────────────────────────────────────────────────────

class ParticipantInfo(BaseModel):
    student_id: int
    full_name: str
    gender: str
    weight_class: Optional[float] = None
    model_config = {"from_attributes": True}


# ── Tournament structure ───────────────────────────────────────────────────────

class WeightClassItem(BaseModel):
    id: int
    weight_class_name: str
    gender: str
    total_players: int
    bracket_status: str
    participants: list[ParticipantInfo] = []


class AgeTypeItem(BaseModel):
    code: str
    description: str
    weight_classes: list[WeightClassItem]


class CategoryItem(BaseModel):
    category: str
    age_types: list[AgeTypeItem]


class TournamentStructure(BaseModel):
    tournament_id: int
    tournament_name: str
    tournament_status: str
    categories: list[CategoryItem]


class TournamentListItem(BaseModel):
    id: int
    name: str
    sport_icon: Optional[str] = None
    status: str
    created_at: datetime
    structure_mode: str
    primary_color: Optional[str] = None
    gradient_primary: Optional[str] = None


class TournamentConfig(BaseModel):
    default_round_duration_seconds: int = 180
    default_break_duration_seconds: int = 30
    default_performance_duration_seconds: int = 120
    consensus_window_secs: float = 1.0
    consensus_min_votes: int = 3
    model_config = {"from_attributes": True}


class TournamentConfigUpdate(BaseModel):
    default_round_duration_seconds: Optional[int] = Field(None, ge=30, le=600)
    default_break_duration_seconds: Optional[int] = Field(None, ge=10, le=300)
    default_performance_duration_seconds: Optional[int] = Field(None, ge=30, le=600)
    consensus_window_secs: Optional[float] = Field(None, ge=0.5, le=5.0)
    consensus_min_votes: Optional[int] = Field(None, ge=1, le=5)


class CreateTournamentRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    sport_icon: Optional[str] = Field(None, max_length=10)
    template_id: Optional[int] = None
    primary_color: Optional[str] = Field(None, max_length=7)
    gradient_primary: Optional[str] = Field(None, max_length=7)


class CreateTournamentResponse(BaseModel):
    id: int
    name: str
    sport_icon: Optional[str] = None
    status: str
    created_at: datetime
    structure_mode: str
    primary_color: Optional[str] = None
    gradient_primary: Optional[str] = None


class UpdateTournamentRequest(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    sport_icon: Optional[str] = Field(None, max_length=10)
    primary_color: Optional[str] = Field(None, max_length=7)
    gradient_primary: Optional[str] = Field(None, max_length=7)


class DeleteTournamentResponse(BaseModel):
    deleted_tournament_id: int


class TournamentTemplateSummary(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    structure_mode: str
    source_tournament_id: Optional[int] = None
    node_count: int
    kata_count: int
    created_at: datetime
    updated_at: datetime


class CreateTournamentTemplateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=500)
    source_tournament_id: int
    copy_katas: bool = True


class CreateTournamentTemplateResponse(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    structure_mode: str
    source_tournament_id: Optional[int] = None
    node_count: int
    kata_count: int
    created_at: datetime
    updated_at: datetime


# ── Bracket match ──────────────────────────────────────────────────────────────

class BracketMatchOut(BaseModel):
    id: int
    round: int
    match_number: int
    match_code: Optional[str] = None
    schedule_order: Optional[int] = None
    player1_name: Optional[str] = None
    player2_name: Optional[str] = None
    score1: Optional[int] = None
    score2: Optional[int] = None
    winner: Optional[int] = None
    status: str
    is_bye: bool = False
    next_match_id: Optional[int] = None
    model_config = {"from_attributes": True}


class BracketOut(BaseModel):
    weight_class_id: int
    weight_class_name: str
    gender: str
    bracket_status: str
    matches: list[BracketMatchOut]


class BracketExportPathOut(BaseModel):
    node_id: int
    node_path: str
    tree_path: str
    path_segments: list[str]
    weight_class_id: int
    weight_class_name: str
    gender: str
    total_players: int
    bracket_status: str
    players: list[str] = []
    matches: list[BracketMatchOut]


class BracketExportOut(BaseModel):
    tournament_id: int
    tournament_name: str
    tournament_status: str
    scope: Literal["single", "all"]
    total_paths: int
    paths: list[BracketExportPathOut]


# ── Match result input ─────────────────────────────────────────────────────────

class MatchResultIn(BaseModel):
    winner: int   # 1 or 2
    score1: int = 0
    score2: int = 0


# ── Generate all matches output ───────────────────────────────────────────────

class GenerateMatchesOut(BaseModel):
    tournament_id: int
    generated_weight_classes: int
    skipped_weight_classes: int
    total_bracket_matches: int
    total_quyen_slots: int
    bye_matches: int


# ── Generate schedule output ──────────────────────────────────────────────────

class GenerateScheduleOut(BaseModel):
    tournament_id: int
    court_a_count: int
    court_b_count: int
    total_scheduled: int


# ── Quyen slot output ─────────────────────────────────────────────────────────

class QuyenSlotOut(BaseModel):
    id: int
    tournament_id: int
    weight_class_id: Optional[int] = None
    weight_class_name: str
    node_id: Optional[int] = None
    node_path: Optional[str] = None
    representative_type: Literal["student", "club"] = "student"
    player_name: str
    player_club: Optional[str] = None
    club_id: Optional[int] = None
    kata_id: Optional[int] = None
    content_name: str
    court: Optional[str] = None
    schedule_order: Optional[int] = None
    status: str
    performance_duration_seconds: int = 120
    started_at: Optional[datetime] = None
    scoring_started_at: Optional[datetime] = None
    scored_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    confirmed_at: Optional[datetime] = None
    official_score: Optional[float] = None
    total_judge_score: Optional[int] = None
    highest_judge_score: Optional[int] = None
    lowest_judge_score: Optional[int] = None
    is_disqualified: bool = False
    assigned_judges_count: int = 0
    ready_judges_count: int = 0
    submitted_judges_count: int = 0
    model_config = {"from_attributes": True}


class QuyenJudgeScoreOut(BaseModel):
    judge_slot: int
    assigned_user_id: Optional[int] = None
    assigned_user_name: Optional[str] = None
    is_ready: bool = False
    ready_at: Optional[datetime] = None
    score: Optional[int] = None
    submitted_at: Optional[datetime] = None


class QuyenRankingItemOut(BaseModel):
    slot_id: int
    rank: int
    player_name: str
    player_club: Optional[str] = None
    official_score: float
    total_judge_score: int
    highest_judge_score: int
    lowest_judge_score: int
    medal: Optional[str] = None


class QuyenRankingGroupOut(BaseModel):
    node_id: Optional[int] = None
    node_path: Optional[str] = None
    content_name: str
    status: Literal["pending", "ready"]
    items: list[QuyenRankingItemOut] = []


class QuyenScoringDetailOut(BaseModel):
    slot: QuyenSlotOut
    tournament_name: str
    tree_path: Optional[str] = None
    judges: list[QuyenJudgeScoreOut]
    ranking_group: Optional[QuyenRankingGroupOut] = None
    audit_count: int = 0


class QuyenJudgePanelOut(BaseModel):
    slot: QuyenSlotOut
    tournament_name: str
    tree_path: Optional[str] = None
    judge: QuyenJudgeScoreOut


class QuyenDisplayJudgeOut(BaseModel):
    judge_slot: int
    assigned_user_name: Optional[str] = None
    is_ready: bool = False
    has_submitted: bool = False
    score: Optional[int] = None


class QuyenDisplayOut(BaseModel):
    slot: QuyenSlotOut
    tournament_name: str
    tree_path: Optional[str] = None
    judges: list[QuyenDisplayJudgeOut]


# ── Extended bracket match for schedule ───────────────────────────────────────

class ScheduleBracketMatchOut(BaseModel):
    id: int
    weight_class_id: int
    weight_class_name: str
    node_path: Optional[str] = None
    gender: str = ""
    category: str = ""
    age_type_code: str = ""
    match_code: Optional[str] = None
    round: int
    round_label: str
    match_number: int
    player1_name: Optional[str] = None
    player2_name: Optional[str] = None
    player1_club: Optional[str] = None
    player2_club: Optional[str] = None
    score1: Optional[int] = None
    score2: Optional[int] = None
    winner: Optional[int] = None
    winner_name: Optional[str] = None
    court: Optional[str] = None
    schedule_order: Optional[int] = None
    status: str
    is_bye: bool = False
    next_match_id: Optional[int] = None
    next_match_code: Optional[str] = None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    assigned_judges_count: int = 0
    ready_judges_count: int = 0
    model_config = {"from_attributes": True}


# ── Schedule summary ──────────────────────────────────────────────────────────

class ScheduleSummary(BaseModel):
    quyen_count: int
    doi_khang_count: int
    ready_count: int
    ongoing_count: int
    scoring_count: int = 0
    scored_count: int = 0
    completed_count: int
    pending_count: int


# ── Full schedule output ──────────────────────────────────────────────────────

class TournamentScheduleOut(BaseModel):
    tournament_id: int
    tournament_name: str
    tournament_status: str
    summary: ScheduleSummary
    quyen_slots: list[QuyenSlotOut]
    bracket_matches: list[ScheduleBracketMatchOut]


# ── Update schedule (drag & drop reorder + court change) ─────────────────────

class ScheduleItemIn(BaseModel):
    id: int
    schedule_order: int
    court: Optional[str] = None

    @field_validator("court")
    @classmethod
    def validate_court(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in ("A", "B"):
            raise ValueError("court phải là 'A' hoặc 'B'")
        return v


class UpdateScheduleIn(BaseModel):
    bracket_matches: list[ScheduleItemIn] = []
    quyen_slots: list[ScheduleItemIn] = []


class UpdateScheduleOut(BaseModel):
    updated_bracket: int
    updated_quyen: int


# ── Medal tally ───────────────────────────────────────────────────────────────

class WeightClassMedal(BaseModel):
    weight_class_id: int
    weight_class_name: str
    gender: str
    tree_path: str = ""
    status: str = "completed"  # "completed" | "in_progress"
    gold: Optional[str] = None
    gold_club: Optional[str] = None
    silver: Optional[str] = None
    silver_club: Optional[str] = None
    bronze: list[str] = []
    bronze_clubs: list[str] = []


class QuyenMedalGroup(BaseModel):
    node_id: Optional[int] = None
    node_path: Optional[str] = None
    content_name: str
    kata_id: Optional[int] = None
    status: str = "completed"  # "completed" | "in_progress"
    gold: Optional[str] = None
    gold_club: Optional[str] = None
    gold_club_id: Optional[int] = None
    silver: Optional[str] = None
    silver_club: Optional[str] = None
    silver_club_id: Optional[int] = None
    bronze: list[str] = []
    bronze_clubs: list[str] = []
    bronze_club_ids: list[Optional[int]] = []


class MedalTallyOut(BaseModel):
    tournament_id: int
    tournament_name: str
    weight_class_medals: list[WeightClassMedal]
    quyen_medals: list[QuyenMedalGroup] = []


class ClubMedalRank(BaseModel):
    rank: int
    club_id: int
    club_name: str
    gold: int
    silver: int
    bronze: int
    total: int
    athlete_count: int


class ClubMedalTallyOut(BaseModel):
    tournament_id: int
    tournament_name: str
    rankings: list[ClubMedalRank]


# ── Match detail (for scoring panel) ──────────────────────────────────────────

class MatchJudgeAssignmentOut(BaseModel):
    judge_slot: int
    assigned_user_id: Optional[int] = None
    assigned_user_name: Optional[str] = None
    is_ready: bool = False
    ready_at: Optional[datetime] = None
    score1: Optional[int] = None
    score2: Optional[int] = None
    has_submitted: bool = False


class MatchJudgePanelOut(BaseModel):
    match_id: int
    tournament_name: str
    player1_name: str | None
    player2_name: str | None
    weight_class_name: str
    status: str
    started_at: Optional[datetime] = None
    judge: MatchJudgeAssignmentOut | None = None


class MatchDetailOut(BaseModel):
    id: int
    tournament_id: int
    match_code: str | None
    round: int
    match_number: int
    court: str | None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    player1_name: str | None
    player2_name: str | None
    player1_club: str | None = None
    player2_club: str | None = None
    player1_avatar_url: str | None = None
    player2_avatar_url: str | None = None
    score1: int | None
    score2: int | None
    winner: int | None
    status: str
    is_bye: bool
    weight_class_name: str
    category: str
    age_type_code: str
    gender: str
    current_hiep: int = 1
    match_phase: str = "not_started"
    round_duration_seconds: int = 180
    break_duration_seconds: int = 30
    assigned_judges_count: int = 0
    ready_judges_count: int = 0
    submitted_judges_count: int = 0
    judges: list[MatchJudgeAssignmentOut] = []
    timer_active: bool = False
    model_config = ConfigDict(from_attributes=True)


class MatchJudgeAssignmentIn(BaseModel):
    judge_slot: int = Field(..., ge=1, le=5)
    user_id: int = Field(..., gt=0)


class SetMatchHiepIn(BaseModel):
    hiep: int = Field(..., ge=1)


class EndMatchRoundIn(BaseModel):
    score1: int
    score2: int


class DrawResultIn(BaseModel):
    winner: int = Field(..., ge=1, le=2)


class MatchConfigIn(BaseModel):
    round_duration_seconds: int | None = Field(None, ge=1)
    break_duration_seconds: int | None = Field(None, ge=1)


class LiveScoreIn(BaseModel):
    score1: int = Field(..., ge=0)
    score2: int = Field(..., ge=0)
    yellow_cards1: int | None = None
    yellow_cards2: int | None = None


class ScoreLogIn(BaseModel):
    action: str = Field(..., max_length=50)
    side: int | None = Field(None, ge=1, le=2)
    delta: int | None = None
    score1_after: int | None = None
    score2_after: int | None = None
    description: str | None = Field(None, max_length=255)


class ScoreLogOut(BaseModel):
    id: int
    match_id: int
    actor_type: str
    actor_name: str | None
    action: str
    side: int | None
    delta: int | None
    score1_after: int | None
    score2_after: int | None
    match_phase: str | None
    description: str | None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class ConsensusTurnVoteOut(BaseModel):
    id: int
    judge_slot: int
    judge_user_id: int | None
    player_side: str
    score_type: str
    press_order: int
    model_config = ConfigDict(from_attributes=True)


class ConsensusTurnOut(BaseModel):
    id: int
    match_id: int
    match_phase: str | None
    is_consensus: bool
    result_side: str | None
    result_type: str | None
    result_delta: int | None
    score1_after: int | None
    score2_after: int | None
    agreeing_slots: str | None
    created_at: datetime
    votes: list[ConsensusTurnVoteOut] = []
    model_config = ConfigDict(from_attributes=True)


class UpdateMatchJudgeSetupIn(BaseModel):
    judges: list[MatchJudgeAssignmentIn]


class MatchJudgeReadyIn(BaseModel):
    ready: bool = True


class MatchJudgeScoreIn(BaseModel):
    score1: int = Field(..., ge=0)
    score2: int = Field(..., ge=0)


class RefereeCurrentAssignmentOut(BaseModel):
    kind: Literal["match", "quyen"]
    entity_id: int
    judge_slot: int
    route: str
    status: str
    title: str


class QuyenJudgeAssignmentIn(BaseModel):
    judge_slot: int = Field(..., ge=1, le=5)
    user_id: int = Field(..., gt=0)


class UpdateQuyenJudgeSetupIn(BaseModel):
    judges: list[QuyenJudgeAssignmentIn]
    performance_duration_seconds: int = Field(..., ge=10, le=600)


class QuyenJudgeReadyIn(BaseModel):
    ready: bool = True


class QuyenJudgeScoreIn(BaseModel):
    score: int = Field(..., ge=80, le=100)

File: frontend/src/types/tournament.ts
export interface ParticipantInfo {
  student_id: number
  full_name: string
  gender: string
  weight_class: number | null
}

export interface WeightClassItem {
  id: number
  weight_class_name: string
  gender: string  // 'M' | 'F'
  total_players: number
  bracket_status: 'NOT_GENERATED' | 'GENERATING' | 'GENERATED'
  players: string[] | null
  participants: ParticipantInfo[]
}

export interface AgeTypeItem {
  code: string
  description: string
  weight_classes: WeightClassItem[]
}

export interface CategoryItem {
  category: string
  age_types: AgeTypeItem[]
}

export interface TournamentStructure {
  tournament_id: number
  tournament_name: string
  tournament_status: string  // 'DRAFT' | 'PUBLISHED' | 'ONGOING' | 'COMPLETED'
  categories: CategoryItem[]
}

export interface TournamentListItem {
  id: number
  name: string
  sport_icon?: string | null
  status: string
  created_at: string
  structure_mode: 'legacy' | 'dynamic'
  primary_color?: string | null
  gradient_primary?: string | null
}

export interface TournamentConfig {
  default_round_duration_seconds: number
  default_break_duration_seconds: number
  default_performance_duration_seconds: number
  consensus_window_secs: number
  consensus_min_votes: number
}

export interface TournamentConfigUpdate {
  default_round_duration_seconds?: number | null
  default_break_duration_seconds?: number | null
  default_performance_duration_seconds?: number | null
  consensus_window_secs?: number | null
  consensus_min_votes?: number | null
}

export interface TournamentTemplateSummary {
  id: number
  name: string
  description: string | null
  structure_mode: 'legacy' | 'dynamic'
  source_tournament_id: number | null
  node_count: number
  kata_count: number
  created_at: string
  updated_at: string
}

export interface BracketMatch {
  id: number
  round: number
  match_number: number
  match_code: string | null
  schedule_order?: number | null
  player1_name: string | null
  player2_name: string | null
  score1: number | null
  score2: number | null
  winner: number | null   // 1 or 2
  status: 'pending' | 'ready' | 'ongoing' | 'completed'
  is_bye: boolean
  next_match_id: number | null
}

export interface BracketOut {
  weight_class_id: number
  weight_class_name: string
  bracket_status: string
  matches: BracketMatch[]
}

export interface BracketExportPath {
  node_id: number
  node_path: string
  tree_path: string
  path_segments: string[]
  weight_class_id: number
  weight_class_name: string
  gender: string
  total_players: number
  bracket_status: 'NOT_GENERATED' | 'GENERATING' | 'GENERATED'
  players: string[]
  matches: BracketMatch[]
}

export interface BracketExportData {
  tournament_id: number
  tournament_name: string
  tournament_status: string
  scope: 'single' | 'all'
  total_paths: number
  paths: BracketExportPath[]
}

export interface MatchResultIn {
  winner: number
  score1: number
  score2: number
}

// ── Schedule types ─────────────────────────────────────────────────────────

export interface QuyenSlot {
  id: number
  tournament_id: number
  weight_class_id: number | null
  weight_class_name: string
  node_id: number | null
  node_path: string | null
  representative_type: 'student' | 'club'
  player_name: string
  player_club: string | null
  club_id: number | null
  kata_id: number | null
  content_name: string
  court: string | null      // 'A' | 'B'
  schedule_order: number | null
  status: 'pending' | 'ready' | 'checking' | 'ongoing' | 'scoring' | 'completed'
  performance_duration_seconds: number
  started_at: string | null
  scoring_started_at: string | null
  scored_at: string | null
  finished_at: string | null
  confirmed_at: string | null
  official_score: number | null
  total_judge_score: number | null
  highest_judge_score: number | null
  lowest_judge_score: number | null
  is_disqualified: boolean
  assigned_judges_count: number
  ready_judges_count: number
  submitted_judges_count: number
}

export interface ScheduleBracketMatch {
  id: number
  weight_class_id: number
  weight_class_name: string
  node_path: string | null
  gender: string
  category: string
  age_type_code: string
  match_code: string | null
  round: number
  round_label: string
  match_number: number
  player1_name: string | null
  player2_name: string | null
  player1_club: string | null
  player2_club: string | null
  score1: number | null
  score2: number | null
  winner: number | null
  winner_name: string | null
  court: string | null
  schedule_order: number | null
  status: 'pending' | 'ready' | 'ongoing' | 'completed'
  is_bye: boolean
  next_match_id: number | null
  next_match_code: string | null
  started_at: string | null
  finished_at: string | null
  assigned_judges_count: number
  ready_judges_count: number
}

export interface ScheduleSummary {
  quyen_count: number
  doi_khang_count: number
  ready_count: number
  ongoing_count: number
  scoring_count: number
  scored_count: number
  completed_count: number
  pending_count: number
}

export interface TournamentSchedule {
  tournament_id: number
  tournament_name: string
  tournament_status: string
  summary: ScheduleSummary
  quyen_slots: QuyenSlot[]
  bracket_matches: ScheduleBracketMatch[]
}

// ── Medal tally ────────────────────────────────────────────────────────────

export interface WeightClassMedal {
  weight_class_id: number
  weight_class_name: string
  gender: string
  tree_path: string
  status: 'completed' | 'in_progress'
  gold: string | null
  gold_club: string | null
  silver: string | null
  silver_club: string | null
  bronze: string[]
  bronze_clubs: string[]
}

export interface QuyenMedalGroup {
  node_id: number | null
  node_path: string | null
  content_name: string
  kata_id: number | null
  status: 'completed' | 'in_progress'
  gold: string | null
  gold_club: string | null
  gold_club_id: number | null
  silver: string | null
  silver_club: string | null
  silver_club_id: number | null
  bronze: string[]
  bronze_clubs: string[]
  bronze_club_ids: (number | null)[]
}

export interface MedalTally {
  tournament_id: number
  tournament_name: string
  weight_class_medals: WeightClassMedal[]
  quyen_medals: QuyenMedalGroup[]
}

export interface ClubMedalRank {
  rank: number
  club_id: number
  club_name: string
  gold: number
  silver: number
  bronze: number
  total: number
  athlete_count: number
}

export interface ClubMedalTally {
  tournament_id: number
  tournament_name: string
  rankings: ClubMedalRank[]
}

export interface QuyenJudgeScore {
  judge_slot: number
  assigned_user_id: number | null
  assigned_user_name: string | null
  is_ready: boolean
  ready_at: string | null
  score: number | null
  submitted_at: string | null
}

export interface QuyenRankingItem {
  slot_id: number
  rank: number
  player_name: string
  player_club: string | null
  official_score: number
  total_judge_score: number
  highest_judge_score: number
  lowest_judge_score: number
  medal: 'gold' | 'silver' | 'bronze' | null
}

export interface QuyenRankingGroup {
  node_id: number | null
  node_path: string | null
  content_name: string
  status: 'pending' | 'ready'
  items: QuyenRankingItem[]
}

export interface QuyenScoringDetail {
  slot: QuyenSlot
  tournament_name: string
  tree_path: string | null
  judges: QuyenJudgeScore[]
  ranking_group: QuyenRankingGroup | null
  audit_count: number
}

export interface QuyenJudgePanel {
  slot: QuyenSlot
  tournament_name: string
  tree_path: string | null
  judge: QuyenJudgeScore
}

export interface QuyenDisplayJudge {
  judge_slot: number
  assigned_user_name: string | null
  is_ready: boolean
  has_submitted: boolean
  score: number | null
}

export interface QuyenDisplayDetail {
  slot: QuyenSlot
  tournament_name: string
  tree_path: string | null
  judges: QuyenDisplayJudge[]
}

// ── Dynamic Tournament Structure types ────────────────────────────────────

export interface TournamentStructureNode {
  id:            number
  tournament_id: number
  parent_id:     number | null
  level:         number        // 0=Gender 1=Category 2=Group 3=WeightClass
  node_type:     'group' | 'weight_class'
  name:          string
  node_code:     string | null // machine-readable code: "M","F","1A","45"...
  rule_json:     Record<string, unknown> | null  // eligibility rules
  sort_order:    number
  student_count: number
  created_at:    string
  updated_at:    string
  children:      TournamentStructureNode[]  // only present in tree format
}

export interface TournamentKata {
  id:            number
  tournament_id: number
  division:      'individual' | 'team'
  name:          string
  description:   string | null
  sort_order:    number
  usage_count:   number
  team_size:     number
  min_team_size: number | null
}

export interface ContestTypeItem {
  type:      'sparring' | 'kata'
  kata_id:   number | null
  kata_name: string | null
}

export interface StudentRegistration {
  student_id:    number
  tournament_id: number
  node_id:       number
  classification_node_id: number | null
  sparring_weight_id: number | null
  node_path:     string
  registered_at: string
  contest_types: ContestTypeItem[]
}

export interface NodeStats {
  total_nodes:          number
  total_weight_classes: number
  total_students:       number
}

export interface NodesResponse {
  tournament_id:      number
  tournament_name:    string
  tournament_status:  string
  nodes:              TournamentStructureNode[]
  stats:              NodeStats
}

export interface KatasResponse {
  tournament_id: number
  katas:         TournamentKata[]
  total:         number
}

export interface TeamKataRegistrationItem {
  node_id: number
  node_path: string
  kata_id: number
  kata_name: string
}

export interface TeamKataRegistrationResponse {
  tournament_id: number
  club_id: number
  registrations: TeamKataRegistrationItem[]
}

export interface TeamKataMemberItem {
  student_id:   number
  student_name: string
  club_name:    string
}

export interface TeamKataMembersResponse {
  tournament_id: number
  club_id:       number
  node_id:       number
  kata_id:       number
  team_size:     number
  members:       TeamKataMemberItem[]
}

export interface NodeStudentItem {
  student_id:         number
  student_name:       string
  assigned_node_id:   number
  assigned_node_name: string
  contest_types:      ContestTypeItem[]
}

export interface NodeStudentsResponse {
  node_id:   number
  node_path: string
  students:  NodeStudentItem[]
  total:     number
}

// ── Dynamic bracket tree ──────────────────────────────────────────────────────

export interface BracketLeafParticipant {
  student_id: number
  full_name: string
}

export interface BracketLeafInfo {
  weight_class_id: number
  bracket_status: 'NOT_GENERATED' | 'GENERATING' | 'GENERATED'
  total_players: number
  players: string[]
  participants: BracketLeafParticipant[]
}

export interface BracketNodeItem {
  id: number
  name: string
  node_type: 'group' | 'weight_class'
  sort_order: number
  leaf_info: BracketLeafInfo | null
  children: BracketNodeItem[]
}

export interface BracketTreeResponse {
  tournament_id: number
  tournament_name: string
  tournament_status: string
  nodes: BracketNodeItem[]
}

// ── Match detail (for scoring panel) ───────────────────────────────────────

export interface MatchDetail {
  id: number
  tournament_id: number
  match_code: string | null
  round: number
  match_number: number
  court: string | null
  started_at: string | null
  finished_at: string | null
  player1_name: string | null
  player2_name: string | null
  player1_club: string | null
  player2_club: string | null
  player1_avatar_url: string | null
  player2_avatar_url: string | null
  score1: number | null
  score2: number | null
  winner: number | null
  status: 'pending' | 'ready' | 'checking' | 'ongoing' | 'completed'
  is_bye: boolean
  weight_class_name: string
  category: string
  age_type_code: string
  gender: string
  current_hiep: number
  match_phase: 'not_started' | 'round_1' | 'break' | 'round_2' | 'extra_time' | 'draw_pending' | 'finished' | 'confirmed'
  round_duration_seconds: number
  break_duration_seconds: number
  assigned_judges_count: number
  ready_judges_count: number
  submitted_judges_count: number
  judges: MatchJudgeAssignment[]
  timer_active: boolean
}

export interface MatchJudgeAssignment {
  judge_slot: number
  assigned_user_id: number | null
  assigned_user_name: string | null
  is_ready: boolean
  ready_at: string | null
  score1: number | null
  score2: number | null
  has_submitted: boolean
}

export interface MatchJudgePanelDetail {
  match_id: number
  tournament_name: string
  player1_name: string | null
  player2_name: string | null
  weight_class_name: string
  status: string
  started_at: string | null
  judge: MatchJudgeAssignment | null
}

export interface RefereeCurrentAssignment {
  kind: 'match' | 'quyen'
  entity_id: number
  judge_slot: number
  route: string
  status: string
  title: string
}

export interface MatchScoreLog {
  id: number
  match_id: number
  actor_type: string
  actor_name: string | null
  action: string
  side: number | null
  delta: number | null
  score1_after: number | null
  score2_after: number | null
  match_phase: string | null
  description: string | null
  created_at: string
}

export interface ConsensusTurnVote {
  id: number
  judge_slot: number
  judge_user_id: number | null
  player_side: string   // 'RED' | 'BLUE'
  score_type: string    // '+1' | '+2' | '-1'
  press_order: number
}

export interface ConsensusTurn {
  id: number
  match_id: number
  match_phase: string | null
  is_consensus: boolean
  result_side: string | null
  result_type: string | null
  result_delta: number | null
  score1_after: number | null
  score2_after: number | null
  agreeing_slots: string | null   // "1,2,3"
  created_at: string
  votes: ConsensusTurnVote[]
}

// ── Realtime WebSocket scoring types ────────────────────────────────────────

export interface ConsensusInfo {
  sequence: [string, string][]
  judge_count: number
  red_delta: number
  blue_delta: number
}

export interface PendingSlot {
  playerSide: 'RED' | 'BLUE'
  scoreType: '+1' | '+2' | '-1'
  slotIndex: number
  judgeCount: number
}

export interface JudgeActivity {
  judgeSlot: number
  playerSide: 'RED' | 'BLUE'
  scoreType: '+1' | '+2' | '-1'
  count: number
  accumulatedDelta: number
  pressDeltas: number[]
}

export interface ConfirmedSlot {
  playerSide: 'RED' | 'BLUE'
  scoreType: '+1' | '+2' | '-1'
  delta: number
}

export type WSInbound =
  | { type: 'snapshot'; match_id: number; status: string; score1: number; score2: number; judge_slot: number | null }
  | { type: 'score_update'; match_id: number; score1: number; score2: number; confirmed_slots: ConfirmedSlot[] }
  | { type: 'pending_update'; match_id: number; pending: PendingSlot[]; judgeInputs: JudgeActivity[] }
  | { type: 'match_state'; match_phase: string; status: string; score1: number; score2: number; timer_active: boolean }
  | { type: 'input_ack'; judge_slot: number; player_side: string; score_type: string }
  | { type: 'judge_ready'; ready_count: number }
  | { type: 'error'; code: string }
  | { type: 'pong' }

export type WSOutbound =
  | { type: 'judge_input'; playerSide: 'RED' | 'BLUE'; scoreType: '+1' | '+2' | '-1'; sequenceIndex: number; createdAt: number }
  | { type: 'ping' }
  | { type: 'admin_cmd'; cmd: 'begin' }
  | { type: 'admin_cmd'; cmd: 'end_round'; score1: number; score2: number }
  | { type: 'admin_cmd'; cmd: 'start_round' }
  | { type: 'admin_cmd'; cmd: 'draw_result'; winner: 1 | 2 }
  | { type: 'admin_cmd'; cmd: 'live_score'; score1: number; score2: number; yellow_cards1?: number | null; yellow_cards2?: number | null }
  | { type: 'admin_cmd'; cmd: 'timer_active'; active: boolean }
